zkevm-circuits/src/evm_circuit/util.rs

```diff
@@ -162,43 +172,88 @@ pub struct StoredExpression<F> {
         offset: usize,
     ) -> Result<AssignedCell<F, F>, Error> {
         let value = self.expr.evaluate(
-            &|scalar| scalar,
+            &|scalar| Value::known(scalar),
             &|_| unimplemented!("selector column"),
             &|fixed_query| {
-                region.get_fixed(offset, fixed_query.column_index(), fixed_query.rotation())
+                Value::known(region.get_fixed(
+                    offset,
+                    fixed_query.column_index(),
+                    fixed_query.rotation(),
+                ))
             },
             &|advide_query| {
-                region.get_advice(offset, advide_query.column_index(), advide_query.rotation())
-            },
-            &|instance_query| {
-                region.get_instance(
+                Value::known(region.get_advice(
                     offset,
-                    instance_query.column_index(),
-                    instance_query.rotation(),
-                )
+                    advide_query.column_index(),
+                    advide_query.rotation(),
+                ))
             },
-            &|_| unimplemented!(),
+            &|_| unimplemented!("instance column"),
+            &|challenge| *region.challenges().indexed()[challenge.index()],
             &|a| -a,
             &|a, b| a + b,
             &|a, b| a * b,
-            &|a, scalar| a * scalar,
+            &|a, scalar| a * Value::known(scalar),
         );
-        self.cell.assign(region, offset, Value::known(value))
+        self.cell.assign(region, offset, value)
     }
 }
 
-#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub(crate) enum CellType {
-    Storage,
+    StoragePhase1,
+    StoragePhase2,
+    StoragePhase3,
     StoragePermutation,
-    Lookup(Table),
+    Lookup(Table)
+}
+
+impl CellType {
+    fn expr_phase<F: FieldExt>(value: &Expression<F>) -> u8 {
+        use Expression::*;
+        match value {
+            Challenge(challenge) => challenge.phase() + 1,
+            Constant(_) | Selector(_) | Fixed(_) | Advice(_) | Instance(_) => 0,
+            Negated(a) | Expression::Scaled(a, _) => Self::expr_phase(a),
+            Sum(a, b) | Product(a, b) => std::cmp::max(Self::expr_phase(a), Self::expr_phase(b)),
+        }
+    }
+
+    /// Return the phase of the expression
+    pub(crate) fn storage_for<F: FieldExt>(value: &Expression<F>) -> CellType {
+        match Self::expr_phase(value) {
+            0 => CellType::StoragePhase1,
+            1 => CellType::StoragePhase2,
+            2 => CellType::StoragePhase3,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Return the phase for the inverse of an the expression
+    pub(crate) fn storage_for_inv<F: FieldExt>(value: &Expression<F>) -> CellType {
+        match Self::expr_phase(value) {
+            0 => CellType::StoragePhase1,
+            1 => CellType::StoragePhase3,
```

Comment:
I can't think about the reason to require phase 3 here? Would you like to explain it more?

Reply:
hmm sorry I think I'm confused about this function now, so what's the difference between `storage_for` and `storage_for_inv`? Inverse of a value should be able to live in the same phase as the value, right?

Reply:
I understand, my question is more like, what makes inverse storage need one extra phase to be stored? For example `advice_at_phase_1 + challenge_usable_after_phase_1` we should be able to store its inverse in phase 2 right? Because this expression is evaluateable in phase 2.